File: php/src/Webhook/Webhook.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Webhook;

use Fivexer\SDK\Exception\FivexerException;

/**
 * Webhook signature verification for the x-fivexer-signature header.
 *
 * Mirrors the platform's signing exactly:
 *   header  : x-fivexer-signature
 *   format  : t=<unix-ms>,v1=<hex64 HMAC-SHA256>
 *   message : "<timestamp>.<raw-body>"
 *   replay  : rejected if |now - timestamp| > tolerance (default +/- 5 min)
 *   compare : hash_equals (constant-time)
 *
 * The body must be the RAW request body bytes - exactly what was received, before
 * any JSON re-serialization, otherwise the HMAC will not match.
 */
final class Webhook
{
    public const SIGNATURE_HEADER = 'x-fivexer-signature';
    public const DEFAULT_TOLERANCE_MS = 5 * 60 * 1000;

    /**
     * Verify the signature and return the parsed event.
     *
     * @param string $payload Raw request body
     * @param string $header The signature header value
     * @param string $secret The webhook secret
     * @param int<0, max>|null $toleranceMs Replay-window tolerance in ms; null = default
     * @param int<0, max>|null $nowMs Current time in ms; null = system clock (for testing)
     * @throws FivexerException on any verification or parse failure
     */
    public static function constructEvent(
        string $payload,
        string $header,
        string $secret,
        ?int $toleranceMs = null,
        ?int $nowMs = null,
    ): WebhookEvent {
        if (!self::verifySignature($payload, $header, $secret, $toleranceMs, $nowMs)) {
            throw new FivexerException('webhook signature verification failed');
        }
        return WebhookEvent::fromPayload($payload);
    }

    /**
     * @return bool true iff the header is well-formed, within tolerance, and the HMAC matches.
     */
    public static function verifySignature(
        string $payload,
        string $header,
        string $secret,
        ?int $toleranceMs = null,
        ?int $nowMs = null,
    ): bool {
        $parsed = self::parseSignatureHeader($header);
        if ($parsed === null) {
            return false;
        }
        [$timestamp, $signatureHex] = $parsed;

        $tolerance = $toleranceMs ?? self::DEFAULT_TOLERANCE_MS;
        $now = $nowMs ?? (int) (\microtime(true) * 1000);
        if (\abs($now - $timestamp) > $tolerance) {
            return false;
        }

        $expected = self::computeSignature($secret, $timestamp, $payload);
        return \hash_equals($expected, $signatureHex);
    }

    /** @return array{0:int<0,max>,1:string}|null [timestampMs, signatureHex64lowercase] or null if malformed. */
    private static function parseSignatureHeader(string $header): ?array
    {
        $header = \trim($header);
        $parts = \explode(',', $header);
        if (\count($parts) !== 2) {
            return null;
        }
        $tPart = \trim($parts[0]);
        $vPart = \trim($parts[1]);
        if (!\str_starts_with($tPart, 't=') || !\str_starts_with($vPart, 'v1=')) {
            return null;
        }
        $tsStr = \substr($tPart, 2);
        $sig = \strtolower(\substr($vPart, 3));
        if ($tsStr === '' || !\ctype_digit($tsStr)) {
            return null;
        }
        // exactly 64 lowercase hex chars (sha256)
        if (\strlen($sig) !== 64 || !\ctype_xdigit($sig)) {
            return null;
        }
        // ctype_digit guarantees a non-negative integer.
        return [(int) $tsStr, $sig];
    }

    private static function computeSignature(string $secret, int $timestamp, string $body): string
    {
        $message = $timestamp . '.' . $body;
        return \hash_hmac('sha256', $message, $secret);
    }
}

File: java/src/main/java/io/fivexer/sdk/webhook/Webhook.java
package io.fivexer.sdk.webhook;

import io.fivexer.sdk.FivexerException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Objects;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Webhook signature verification for the {@code x-fivexer-signature} header.
 *
 * <p>Mirrors the platform's signing exactly:
 * <pre>
 *   header  : x-fivexer-signature
 *   format  : t=&lt;unix-ms&gt;,v1=&lt;hex64 HMAC-SHA256&gt;
 *   message : "&lt;timestamp&gt;.&lt;raw-body&gt;"
 *   replay  : rejected if |now - timestamp| &gt; tolerance (default ±5 min)
 *   compare : constant-time (MessageDigest.isEqual)
 * </pre>
 *
 * <p>The body must be the <b>raw</b> request body bytes — exactly what was received, before
 * any JSON re-serialization, otherwise the HMAC will not match.
 */
public final class Webhook {

    public static final String SIGNATURE_HEADER = "x-fivexer-signature";
    public static final long DEFAULT_TOLERANCE_MS = 5L * 60 * 1000;

    private Webhook() {}

    /** Verify the signature and return the parsed event. Throws {@link FivexerException} on any failure. */
    public static WebhookEvent constructEvent(byte[] payload, String header, String secret) {
        return constructEvent(payload, header, secret, DEFAULT_TOLERANCE_MS, System.currentTimeMillis());
    }

    public static WebhookEvent constructEvent(byte[] payload, String header, String secret, long toleranceMs) {
        return constructEvent(payload, header, secret, toleranceMs, System.currentTimeMillis());
    }

    public static WebhookEvent constructEvent(byte[] payload, String header, String secret, long toleranceMs, long nowMs) {
        if (!verifySignature(payload, header, secret, toleranceMs, nowMs)) {
            throw new FivexerException("webhook signature verification failed");
        }
        return WebhookEvent.fromPayload(payload);
    }

    /** @return {@code true} iff the header is well-formed, within tolerance, and the HMAC matches. */
    public static boolean verifySignature(byte[] payload, String header, String secret, long toleranceMs, long nowMs) {
        Objects.requireNonNull(payload, "payload");
        ParsedSig parsed = parseSignatureHeader(header);
        if (parsed == null) {
            return false;
        }
        if (Math.abs(nowMs - parsed.timestamp) > toleranceMs) {
            return false;
        }
        byte[] expected = computeSignature(secret, parsed.timestamp, payload);
        byte[] supplied = parsed.signatureBytes;
        // parseSignatureHeader guarantees exactly 64 hex chars → 32 bytes, matching expected;
        // MessageDigest.isEqual is constant-time and also guards length.
        return MessageDigest.isEqual(expected, supplied);
    }

    // ---- internals ----

    private static byte[] computeSignature(String secret, long timestamp, byte[] body) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            mac.update(Long.toString(timestamp).getBytes(StandardCharsets.US_ASCII));
            mac.update((byte) '.');
            mac.update(body);
            return mac.doFinal();
        } catch (Exception e) {
            throw new FivexerException("failed to compute webhook signature", e);
        }
    }

    /** Parsed {@code t=<digits>,v1=<64 hex>} into timestamp + raw signature bytes. */
    private static final class ParsedSig {
        final long timestamp;
        final byte[] signatureBytes;

        ParsedSig(long timestamp, byte[] signatureBytes) {
            this.timestamp = timestamp;
            this.signatureBytes = signatureBytes;
        }
    }

    private static ParsedSig parseSignatureHeader(String header) {
        if (header == null) {
            return null;
        }
        String trimmed = header.trim();
        // limit -1 preserves trailing empty strings so Python parity holds for malformed headers
        String[] parts = trimmed.split(",", -1);
        if (parts.length != 2) {
            return null;
        }
        String tPart = parts[0].trim();
        String vPart = parts[1].trim();
        if (!tPart.startsWith("t=") || !vPart.startsWith("v1=")) {
            return null;
        }
        String tsStr = tPart.substring(2);
        String sig = vPart.substring(3);
        for (int i = 0; i < tsStr.length(); i++) {
            if (!Character.isDigit(tsStr.charAt(i))) {
                return null;
            }
        }
        if (tsStr.isEmpty() || sig.length() != 64 || !isHex(sig)) {
            return null;
        }
        byte[] sigBytes = hexToBytes(sig.toLowerCase());
        return new ParsedSig(Long.parseLong(tsStr), sigBytes);
    }

    private static boolean isHex(String s) {
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'))) {
                return false;
            }
        }
        return true;
    }

    private static byte[] hexToBytes(String hex) {
        byte[] out = new byte[hex.length() / 2];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }
}
